docs/virtualpage.rst

```diff
@@ -0,0 +1,293 @@
+
+    $t = $layout->add('Tabs');
+
+    $t->addTab('Tab1')->add('LoremIpsum'); // regular tab
+    $t->addTab('Tab2', function($p){ $p->add('LoremIpsum'); }); // dynamic tab
+
+uses the following implementation of addTab()::
+
+    public function addTab($name = null, $action = null)
+    {
+        // add tabs menu item
+        $item = $this->add([$item, 'class'=>['item']], 'Menu');
+        $item->setElement('a');
+        $item->setAttr('data-tab', $item->name);
+
+        // add tabs sub-view
+        $sub = $this->add(['View', 'class'=>['ui tab']], 'Tabs');
+        $sub->setAttr('data-tab', $item->name);
+
+        // if there is callback action, then
+        if ($action && is_callable($action)) {
+            $vp = $sub->add('VirtualPage');
+            $item->setPath($vp->getUrl());
+
+            $vp->set($action);
+        }
+
+        return  $sub;
+    }
+
+You do not need to fully understand the code to notice that Dynamic Content support requires just
+few lines of extra code.
+
+.. php:method:: getURL($html_wrapping)
+
+    You can use this shortcut method instead of $vp->cb->getURL().
+
+.. php:attr:: $ui
+
+When using 'popup' mode, the output appears inside a `<div class="ui container">`. If you want to change this
+class, you can set $ui property to something else. Try::
+
+    $vp = $layout->add('VirtualPage');
+    $vp->add('LoremIpsum');
+    $vp->ui = 'red inverted segment';
+
+    $label = $layout->add('Label');
+
+    $label->detail = $vp->cb->getURL('popup');
+    $label->link($vp->cb->getURL('popup'));
+
+
+
+
+
+Loader
+------
+
+.. php:class:: Loader
+
+.. php:method:: set()
+
+Loader is designed to delay some slow-loading content by loading it dynamically, after main
+page is rendered.
+
+Comparing to VirtualPage which is a D.Y.I. solution - Loader can be used out of the box.
+Loader extends VirtualPage and is quite similar to it.
+
+Like with a VirtualPage - you should use `set()` to define content that will be loaded dynamically,
+while a spinner is shown to a user::
+
+    $loader = $app->add('Loader');
+    $loader->set(function($p) {
+
+        // Simulate slow-loading component
+        sleep(2);
+        $p->add('LoremIpsum');
+
+    });
+
+A good use-case example would be a dashboard graph. Unlike VirtualPage which is not visible to a regular render,
+Loader needs to occupy some space, so adding other objects inside content can be a good filler until content is loaded::
+
+    $loader = $app->add('Loader');
+    $loader->add(['Message', 'Please wait until we load LoremIpsum...', 'red']);
+    $loader->set(function($p) {
+
+        // Simulate slow-loading component
+        sleep(2);
+        $p->add('LoremIpsum');
+
+    });
+
+
+Triggering Loader
+^^^^^^^^^^^^^^^^^
+
+By default, Loader will display a spinner and will start loading it's contents as soon as DOM Ready() event fires.
+Sometimes you want to control the event. 
+
```

Comment:
Loader in does not start on page load by default. You need to set it with:

```
$loader->needOnPageLoad = true;
```
But we can change the behaviour if you want.

Reply:
I personally think that it should be `true` by default, but of course that depends on use-case.

Reply:
i have read your implementation very througfully, the wrote documentation which diverges slightly from your implementation. Those differences are there due to

 - better argument naming
 - more usable
 - better out of the box experience.

I suggest to adopt documented spec 100%. In this particular case, the documentation explains that the loader will load itself by default automatically as Imants have also suggested, but if you do invoke jsLoad(), that will disable that. My suggestion is to add check inside render() and track call to isLoad through a protected property.